embulk-core/src/main/java/org/embulk/plugin/maven/MavenPluginSource.java

```diff
@@ -0,0 +1,168 @@
+    {
+        final String category;
+        if (InputPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "input";
+        } else if (OutputPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "output";
+        } else if (ParserPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "parser";
+        } else if (FormatterPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "formatter";
+        } else if (DecoderPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "decoder";
+        } else if (EncoderPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "encoder";
+        } else if (FilterPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "filter";
+        } else if (GuessPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "guess";
+        } else if (ExecutorPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "executor";
+        } else {
+            // unsupported plugin category
+            throw new PluginSourceNotMatchException("Plugin interface " + pluginInterface + " is not supported.");
+        }
+
+        if (pluginType.getSourceType() != PluginSource.Type.MAVEN) {
+            throw new PluginSourceNotMatchException();
+        }
+        final MavenPluginType mavenPluginType = (MavenPluginType) pluginType;
+
+        final PluginClassLoaderFactory pluginClassLoaderFactory =
+            this.injector.getInstance(PluginClassLoaderFactory.class);
+
+        final MavenArtifactFinder mavenArtifactFinder;
+        try {
+            mavenArtifactFinder = MavenArtifactFinder.create(getLocalMavenRepository());
+        }
+        catch (MavenRepositoryNotFoundException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        final Path jarPath;
+        try {
+            jarPath = mavenArtifactFinder.findMavenArtifactJar(
+                mavenPluginType.getGroup(),
+                "embulk-" + category + "-" + mavenPluginType.getName(),
+                mavenPluginType.getVersion());
+        }
+        catch (MavenArtifactNotFoundException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        final Class pluginMainClass;
+        try (JarPluginLoader loader = JarPluginLoader.load(jarPath, pluginClassLoaderFactory)) {
+            pluginMainClass = loader.getPluginMainClass();
+        }
+        catch (InvalidJarPluginException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        if (!pluginInterface.isAssignableFrom(pluginMainClass)) {
+            throw new PluginSourceNotMatchException(
+                "Plugin class \"" + pluginMainClass.getName() + "\" is not a valid " + category + " plugin.");
+        }
+
+        final Object pluginMainObject;
+        try {
+            pluginMainObject = pluginMainClass.newInstance();
```

Comment:
Since all plugins don't have non-argument constructor, it's better to use `injector.newInstance`. And I think that it needs to wrap the instance with `FileInputRunner` or `FileOutputRunner` if the plugin is based on file based plugin. It seems that `InjectedPluginSource.registerPluginTo` method implementation can be helpful.